### DISCUSSION TITLE
Add support for URL validator options to URLField

### DISCUSSION THREAD

**Discussion by psjamesh** [encode/django-rest-framework#8895] [General] [answered]
It would be useful it the `URLField` class supported passing options to the underlying `URLValidator`.

For example, I would like to verify that a URL is HTTPS only. It would be helpful to define this in the field, eg:

```
url = URLField(url_validator_options={"schemes": ["https"]})
```

This would be a very simple change, adding flexibility to the field to avoid including extra validation elsewhere in the application.

```
class URLField(CharField):
    default_error_messages = {"invalid": _("Enter a valid URL.")}

    def __init__(self, url_validator_options: dict = {}, **kwargs):
        super().__init__(**kwargs)

        validator = URLValidator(
            message=self.error_messages["invalid"], **url_validator_options
        )

        self.validators.append(validator)
```

**Comment by nachollica** [comment, marked as answer]
For your particular example (allow only HTTPS) there's a clean and simple solution that doesn't require modifications in the `URLField` class:

```python
from django.core.validators import URLValidator
from rest_framework import serializers

class MySerializer(serializers.Serializer):
    some_url = serializers.URLField(
        validators=[URLValidator(schemes=["https"])],
    )

assert not MySerializer(data={"some_url": "http://example.com"}).is_valid()
assert MySerializer(data={"some_url": "https://example.com"}).is_valid()
```

Nevertheless, you made me realize that this works because your schemes of interest are a subset of the default schemes of `URLValidator`. But there's something interesting: If someone needs to validate, say, `s3://bucket/prefix` the above code will never work, because both validators (the default one and the explicitly set in `MySerializer`) will run, and the sets of schemes are disjoint.

I propose this answer because it may help in your use case, and I also give +1 because the feature seems to be useful in other scenarios.

**Reply by psjamesh** [reply]
Thanks Ignacio, that's a good solution to my problem.
